### ISSUE TITLE
Version check not working: {"errors":["invalid game"]}

### ISSUE THREAD

**Issue by cpudreams** [itchio/itch.io#496]
Using chrome:

https://itch.io/api/1/x/wharf/latest?target=cpudreams/hunternet&channel_name=windows-alpha

Returns:

{"latest":"926"}

Making a GET request for the same URL from within Unity always returns:

{"errors":["invalid game"]}

I've tried three different methods: Unity's WWW class, System.Net.WebClient and System.Net.WebRequest all with the same result.

Here is my test code:

using UnityEngine;
using System.Collections;
using System.Net;
using System.IO;
using System;

public class ItchIoTest : MonoBehaviour {

  void Start () {
    StartCoroutine(GetVersion());
  }

  IEnumerator GetVersion() {
    var url = @"https://itch.io/api/1/x/wharf/latest?target=cpudreams/hunternet&channel_name=windows-alpha";
    Debug.Log("raw: " + url);

    var builder = new UriBuilder();
    builder.Scheme = "https";
    builder.Host = "itch.io";
    builder.Path = "api/1/x/wharf/latest";
    //builder.Query = "target=cpudreams/hunternet&channel_name=windows-alpha";
    builder.Query = string.Format("{0}={1}&{2}={3}",
                                  WWW.EscapeURL("target"),
                                  WWW.EscapeURL("cpudreams/hunternet"),
                                  WWW.EscapeURL("channel_name"),
                                  WWW.EscapeURL("windows-alpha"));
    Debug.Log("builder: " + builder.ToString());
    var uri = builder.Uri;

    var www = new WWW(url);
    yield return www;  // Wait for download to complete
    Debug.Log("text1: " + www.text);

    WebClient wc = new WebClient();
    wc.Proxy = null;
    Debug.Log("text2: " + wc.DownloadString(uri));

    WebRequest req = WebRequest.Create(uri);
    req.Proxy = null;
    using (WebResponse res = req.GetResponse())
      using (Stream s = res.GetResponseStream())
        using (StreamReader sr = new StreamReader(s))
          Debug.Log("text3: " + sr.ReadToEnd());
  }
}

If the error message included the invalid game name, maybe it would make the source of the problem more obvious.

**Comment by fasterthanlime** [comment]
The reason is that, in Chrome, you're logged into itch.io as the user who owns that game. Since the game is in visibility 'restricted', an anonymous user cannot fetch that data.

I originally had it more lenient but I think it was @leafo's decision to move to this behavior? I'm not sure what the proper way should be for games in private beta (incl. page in restricted mode, invisible to the public) that want to check their version number.

**Comment by cpudreams** [comment]
I think there needs to be some way to access the version number even if visibility is restricted, otherwise, this will get in the way of testing. 

If you are concerned about leaking version numbers maybe add an option to the control panel to "allow version requests in restricted" (or even "always allow version number requests"). It could be off by default and if instead of "invalid game" we got an error message with a link to a support page or FAQ entry to turn on "allow version request in restricted". 

Another option is to allow the version number request if you add the restricted page password to the GET request.

Either way, I think it really needs a more meaningful error message.

Thanks!

**Comment by fasterthanlime** [comment]
> Either way, I think it really needs a more meaningful error message.

I agree with this, the API as it currently stands is suitable for internal usage but falls short in some aspects for external users. Improving that is definitely in our plans.

> I think there needs to be some way to access the version number even if visibility is restricted, otherwise, this will get in the way of testing.

I agree, let's see which way @leafo thinks is best.

**Comment by leafo** [comment]
I think what makes the most sense is to let you generate a api key for your script that only has permission to view the game. That way you can revoke it if necessary.

There currently isn't a way to set scopes on api keys from the web interface though.

**Comment by cpudreams** [comment]
Could you support adding the games password to the query string as a workaround until you add support for API keys? It seems like that would avoid any security concerns since if you have access to the password you already have access to the game's restricted page.

**Comment by fasterthanlime** [comment]
@cpudreams there's also visibility levels that let only people with a download key (& its developers / game admins) install the game, so that wouldn't cover all use cases exactly.

@leafo Instead of cluttering the web interface, how about adding API support for this? Then I can just add a butler command to generate a read-only scoped API key, and eventually expose that via the itch app's UI when I finally code that butler front-end :)

**Comment by cpudreams** [comment]
Sounds like a good plan. In the meantime, I'm just going to fake it by uploading the current version number in the same format as your API to a URL on my domain and then have the game check that instead of the itch API until support is available for API keys. Thanks :)!

**Comment by primaerfunktion** [comment]
Heya! I'd like to resurrect this one, as I also just encountered this. (Honestly a bit surprised there aren't any more responses.) Would be very useful to use it also for private testing.

**Comment by fasterthanlime** [comment]
@primaerfunktion My 2019 thoughts are still aligned with my 2016 thoughts on this, there's been no developments that I'm aware of.

**Comment by ha-weng** [comment]
I'd like to resurrect this yet again since I just realized that I won't be able to test my implementation of the version check before making my project public.
There's also a duplicate of this issue here: https://github.com/itchio/itch.io/issues/560
